CIP-XXXX/README.md

```diff
@@ -0,0 +1,270 @@
+This CIP however doesn't require that that be implemented.
+
+The argument for privacy doesn't apply, private smart contracts can be achieved through
+the use of non-interactive zero-knowledge probabilistic proofs.
+
+### Reference scripts
+
+Currently, different instances of the same script will need their own reference inputs
+since their hashes are different. It seems feasible to allow sharing of a single reference script,
+given the parameters and language version as witnesses, but given the complexity
+involved, it is not specified in this CIP.
+
+### Staking
+
+This makes staking validators much more powerful, since a single protocol can
+now manage many rewards accounts (by instantiating the script with a numeric identifier).
+However, it is arguably not the optimal solution due to the reference
+script problem described above. Even if the reference script problem
+is solved as described above, it seems logical to allow supplying a datum
+to a staking validator, or somehow combining the payment address and staking address for scripts,
+and using the same datum for both, while somehow solving the separate accounts problem.
+
+Given the heavy complexity of fixing staking validators, Merkelising script hashing seems much more feasible.
+
+### Alternatives
+
+#### Parameterised Reference Scripts
+
+See https://github.com/cardano-foundation/CIPs/pull/354.
+
+Seemingly, Merkelisation is a less invasive and possibly cleaner change.
+
+#### Changing how constants are hashed
+
+The hashing of constants might not have a clear best solution, but most notably,
+it is not clear how much/less to Merkelise the hashing.
+E.g., the hashing of data itself could be Merkelised. This is not done in this CIP.
+The hashing of a `Data` constant could also prepend the prefix directly to the serialisation,
+rather than to the hash of the `Data`. It is not clear what is best.
+
+##### Hashing strings, lists, pairs differently
+
+Strings are not very useful in Plutus.
+Hence, the hashing algorithm for them isn't optimised such that
+they can be easily verified.
+
+Strings have essentially no purpose on-chain, since they're only used
+for tracing, which should not be used in production.
+
+In the context of checking applied parameters, it is likely that only
+`Data`, `Integer`, `Bool`, `ByteString`, will be used as parameters,
+since they cover all useful behaviour in an efficient way.
+If you want to parameterise your script by a pair of integers,
+it is likely best to unwrap that into two separate integer parameters
+for the sake of efficiency of _running_ the script, which is likely
+to be more common that checking the parameters.
+
+Built-in lists and pairs are not commonly used as parameters, but it's plausible
+that they might still be the most efficient method in some scenarios.
+Hence, they have been included.
+They use Merkle-tree hashing since that's the simplest and most useful in this case.
+
+## Path to Active
```

Comment:
I think this should have some Acceptance Criteria a la the new CIP-001. Perhaps:
- The ledger specification is updated as necessary
- The Plutus Core specification is updated with the new hashing scheme
- Performance assessment has been performed
- Necessary hashing builtins have been added to PLC and costed
- Example usecases have been validated to run in an acceptable amount of budget considering the increased use of hashing builtins

Reply:
`original` and `script` are constants here. `original` is a fixed script, and can thus be computed beforehand and inlined into the script. `script` comes from `ScriptContext`.